In [2]:
"""
Clustering

The objective function is

f(S,C) = (1/2) ∑x∈S¹ ∑y∈S¹,x≠y δ(x,y)

Example:

Suppose we have one cluster containing three points:

A = (0,0)

B = (2,0)

C = (4,0)

Since there is only one cluster (k = 1), we only evaluate this cluster.

Start with x = A:

(1/2)[δ(A,B) + δ(A,C) + ...]

= (1/2)[2 + 4 + ...]

Running total = 6

Next, let x = B:

(1/2)[2 + 4 + δ(B,A) + δ(B,C) + ...]

= (1/2)[2 + 4 + 2 + 2 + ...]

Running total = 10

Finally, let x = C:

(1/2)[2 + 4 + 2 + 2 + δ(C,A) + δ(C,B)]

= (1/2)[2 + 4 + 2 + 2 + 4 + 2]

= (1/2)(16)

= 8

Notice that every pair is counted twice:

δ(A,B) = δ(B,A)

δ(A,C) = δ(C,A)

δ(B,C) = δ(C,B)

The factor of 1/2 removes this duplicate counting.

Therefore, the objective function is equivalent to summing each unique pair once:

δ(A,B) + δ(A,C) + δ(B,C)

= 2 + 4 + 2

= 8
"""

'\nClustering\n\nThe objective function is\n\nf(S,C) = (1/2) ∑x∈S¹ ∑y∈S¹,x≠y δ(x,y)\n\nExample:\n\nSuppose we have one cluster containing three points:\n\nA = (0,0)\n\nB = (2,0)\n\nC = (4,0)\n\nSince there is only one cluster (k = 1), we only evaluate this cluster.\n\nStart with x = A:\n\n(1/2)[δ(A,B) + δ(A,C) + ...]\n\n= (1/2)[2 + 4 + ...]\n\nRunning total = 6\n\nNext, let x = B:\n\n(1/2)[2 + 4 + δ(B,A) + δ(B,C) + ...]\n\n= (1/2)[2 + 4 + 2 + 2 + ...]\n\nRunning total = 10\n\nFinally, let x = C:\n\n(1/2)[2 + 4 + 2 + 2 + δ(C,A) + δ(C,B)]\n\n= (1/2)[2 + 4 + 2 + 2 + 4 + 2]\n\n= (1/2)(16)\n\n= 8\n\nNotice that every pair is counted twice:\n\nδ(A,B) = δ(B,A)\n\nδ(A,C) = δ(C,A)\n\nδ(B,C) = δ(C,B)\n\nThe factor of 1/2 removes this duplicate counting.\n\nTherefore, the objective function is equivalent to summing each unique pair once:\n\nδ(A,B) + δ(A,C) + δ(B,C)\n\n= 2 + 4 + 2\n\n= 8\n'

In [3]:
"""
When the input contains many repeated points, choosing the initial centers uniformly at random may select multiple centers
that are identical or very close together. This produces poor initial clusters because several centers represent the same region of the data.

A better initialization procedure is to choose the first center uniformly at random from all points.
Then, for each remaining point, compute its distance to the nearest center that has already been chosen.
Select the next center with probability proportional to the square of this distance.
Consequently, points that are far away from all existing centers are much more likely to be selected,
while duplicate or nearby points have a very small probability of being chosen. Repeat this process until k centers have been selected.

This initialization spreads the centers throughout the data, maximizes the number of distant centers chosen,
and typically leads to better clustering results and faster convergence of Lloyd's algorithm.
"""

"\nWhen the input contains many repeated points, choosing the initial centers uniformly at random may select multiple centers\nthat are identical or very close together. This produces poor initial clusters because several centers represent the same region of the data.\n\nA better initialization procedure is to choose the first center uniformly at random from all points.\nThen, for each remaining point, compute its distance to the nearest center that has already been chosen.\nSelect the next center with probability proportional to the square of this distance.\nConsequently, points that are far away from all existing centers are much more likely to be selected,\nwhile duplicate or nearby points have a very small probability of being chosen. Repeat this process until k centers have been selected.\n\nThis initialization spreads the centers throughout the data, maximizes the number of distant centers chosen,\nand typically leads to better clustering results and faster convergence of Lloyd's

In [4]:
"""
Show how to find an optimal k-clustering in polynomial time when there is just one attribute (d = 1):
(points are on a line vs a plane)

The reason this works is that in one dimension every optimal cluster is a contiguous block of sorted points.
The dynamic program simply tries every possible location where the last cluster could begin and chooses
the split that minimizes the total within-cluster variance. This property does not hold in higher dimensions,
which is why the general k-means problem is much harder.

Algorithm:

1. Sort the n points in increasing order.

   x₁ ≤ x₂ ≤ ... ≤ xₙ

2. Precompute the clustering cost for every interval
   [i, j].

   Since every cluster consists of consecutive points,
   compute the cost of placing points
   xᵢ, xᵢ₊₁, ..., xⱼ
   into a single cluster.

   The optimal center is the mean of these points, so

   cost(i,j) =
   Σₜ₌ᵢʲ (xₜ − μᵢⱼ)²

   where

   μᵢⱼ = (xᵢ + ... + xⱼ)/(j−i+1)

3. Let DP[i][m] be the minimum clustering cost for the
   first i sorted points using exactly m clusters.

4. The recurrence is

   DP[i][m] =
   minₜ<m≤i
   { DP[t][m−1] + cost(t+1,i) }

   This means:

   - The first t points are optimally partitioned into
     m−1 clusters.

   - The last cluster contains points
     xₜ₊₁ through xᵢ.

5. Fill the DP table for all values of i and m.

6. The answer is DP[n][k].
"""

'\nShow how to find an optimal k-clustering in polynomial time when there is just one attribute (d = 1):\n(points are on a line vs a plane)\n\nThe reason this works is that in one dimension every optimal cluster is a contiguous block of sorted points.\nThe dynamic program simply tries every possible location where the last cluster could begin and chooses\nthe split that minimizes the total within-cluster variance. This property does not hold in higher dimensions,\nwhich is why the general k-means problem is much harder.\n\nAlgorithm:\n\n1. Sort the n points in increasing order.\n\n   x₁ ≤ x₂ ≤ ... ≤ xₙ\n\n2. Precompute the clustering cost for every interval\n   [i, j].\n\n   Since every cluster consists of consecutive points,\n   compute the cost of placing points\n   xᵢ, xᵢ₊₁, ..., xⱼ\n   into a single cluster.\n\n   The optimal center is the mean of these points, so\n\n   cost(i,j) =\n   Σₜ₌ᵢʲ (xₜ − μᵢⱼ)²\n\n   where\n\n   μᵢⱼ = (xᵢ + ... + xⱼ)/(j−i+1)\n\n3. Let DP[i][m] be the minim

In [5]:
def weighted_majority(experts, predictions, outcomes, gamma):
    n = len(experts)
    T = len(outcomes)
    # Initialize every expert with weight 1
    weights = [1 for _ in range(n)]
    final_predictions = []

    for t in range(T):
        # Experts make predictions for round t
        # predictions[t][i] = prediction of expert i
        q = predictions[t]

        # Calculate weighted votes
        upweight = 0
        downweight = 0

        for i in range(n):
            if q[i] == 1:
                upweight += weights[i]
            else:
                downweight += weights[i]

        # Majority prediction
        if upweight >= downweight:
            prediction = 1
        else:
            prediction = 0

        final_predictions.append(prediction)

        # Actual outcome is revealed
        outcome = outcomes[t]

        # Update weights
        for i in range(n):
            if q[i] != outcome:
                weights[i] *= (1 - gamma)

    return final_predictions

In [7]:
"""
Weighted Majority Algorithm Example

Suppose we have 3 experts:
Expert A
Expert B
Expert C

The parameter is:
gamma = 0.5

This means that when an expert makes a mistake, their weight is multiplied by:

(1 - gamma) = 0.5

Initially all experts have the same weight:

A = 1
B = 1
C = 1

The experts make predictions over 4 rounds:

Round       A    B    C      Actual Outcome

1           1    1    0          1

2           1    0    0          0

3           0    1    1          1

4           1    1    0          1

--------------------------------------------------

Round 1:

Predictions:
A predicts 1
B predicts 1
C predicts 0


Current weights:
A = 1
B = 1
C = 1

Weighted votes:

Prediction 1:
A + B = 1 + 1 = 2

Prediction 0:
C = 1

The algorithm predicts:
p(1) = 1

Actual outcome:
o(1) = 1

Correct experts:
A and B

Wrong expert:
C

Update weights:
A = 1
B = 1
C = 1 * (1 - 0.5)
C = 0.5

New weights:
A = 1
B = 1
C = 0.5

--------------------------------------------------

Round 2:
Predictions:
A predicts 1
B predicts 0
C predicts 0

Current weights:
A = 1
B = 1
C = 0.5

Weighted votes:
Prediction 1:
A = 1

Prediction 0:
B + C = 1 + 0.5 = 1.5

The algorithm predicts:
p(2) = 0

Actual outcome:
o(2) = 0

Correct experts:
B and C

Wrong expert:
A

Update weights:
A = 1 * (1 - 0.5)
A = 0.5

B = 1
C = 0.5

New weights:
A = 0.5
B = 1
C = 0.5

--------------------------------------------------

Round 3:

Predictions:
A predicts 0
B predicts 1
C predicts 1

Current weights:
A = 0.5
B = 1
C = 0.5

Weighted votes:

Prediction 1:
B + C = 1 + 0.5 = 1.5

Prediction 0:
A = 0.5

The algorithm predicts:
p(3) = 1

Actual outcome:
o(3) = 1

Correct experts:
B and C

Wrong expert:
A

Update weights:
A = 0.5 * (1 - 0.5)
A = 0.25

B = 1
C = 0.5

New weights:
A = 0.25
B = 1
C = 0.5

--------------------------------------------------

Round 4:

Predictions:
A predicts 1
B predicts 1
C predicts 0

Current weights:
A = 0.25
B = 1
C = 0.5

Weighted votes:
Prediction 1:
A + B = 0.25 + 1 = 1.25

Prediction 0:
C = 0.5

The algorithm predicts:
p(4) = 1

Actual outcome:
o(4) = 1

Correct experts:
A and B

Wrong expert:
C

Update weights:
A = 0.25
B = 1
C = 0.5 * (1 - 0.5)
C = 0.25

Final weights:
A = 0.25
B = 1
C = 0.25

Conclusion:

The Weighted Majority Algorithm starts by giving every expert equal influence. As the rounds progress,
4experts that make incorrect predictions repeatedly lose weight. Experts that consistently predict correctly
keep their weight and become more influential in the final decision.

The algorithm does not know which expert is best initially. It learns by reducing the influence of experts that make mistakes.
"""

'\nWeighted Majority Algorithm Example\n\nSuppose we have 3 experts:\nExpert A\nExpert B\nExpert C\n\nThe parameter is:\ngamma = 0.5\n\nThis means that when an expert makes a mistake, their weight is multiplied by:\n\n(1 - gamma) = 0.5\n\nInitially all experts have the same weight:\n\nA = 1\nB = 1\nC = 1\n\nThe experts make predictions over 4 rounds:\n\nRound       A    B    C      Actual Outcome\n\n1           1    1    0          1\n\n2           1    0    0          0\n\n3           0    1    1          1\n\n4           1    1    0          1\n\n--------------------------------------------------\n\nRound 1:\n\nPredictions:\nA predicts 1\nB predicts 1\nC predicts 0\n\n\nCurrent weights:\nA = 1\nB = 1\nC = 1\n\nWeighted votes:\n\nPrediction 1:\nA + B = 1 + 1 = 2\n\nPrediction 0:\nC = 1\n\nThe algorithm predicts:\np(1) = 1\n\nActual outcome:\no(1) = 1\n\nCorrect experts:\nA and B\n\nWrong expert:\nC\n\nUpdate weights:\nA = 1\nB = 1\nC = 1 * (1 - 0.5)\nC = 0.5\n\nNew weights:\nA = 1\nB 

In [8]:
"""
gradient of x^2 (derivative)
"""
def f_grad(x):
    return 2 * x

In [9]:
"""
eta: learning rate
"""
def gradient_descent(f_grad, x0, eta, T):
    x = [0] * T
    x[0] = x0
    total = x0

    for t in range(T - 1):
        x[t + 1] = x[t] - eta * f_grad(x[t])
        total += x[t + 1]

    x_avg = total / T

    return x, x_avg

In [11]:
def project(x):
    if x < 0:
        return 0
    elif x > 2:
        return 2
    else:
        return x

In [10]:
"""
Ordinary gradient descent only asks:
"Where is the function decreasing?"

Projected gradient descent asks:
"Where is the function decreasing while staying inside the allowed region?"
"""
def gradient_descent_constrained(f_grad, x0, eta, T, project):
    x = [0] * T
    x[0] = x0
    total = x0

    for t in range(T - 1):
        # Unconstrained gradient step
        x_temp = x[t] - eta * f_grad(x[t])

        # Project back into the feasible set K
        x[t + 1] = project(x_temp)
        total += x[t + 1]

    x_avg = total / T

    return x, x_avg

In [12]:
points, average = gradient_descent_constrained(
    f_grad=f_grad,
    x0=1.5,
    eta=1,
    T=5,
    project=project
)

print("Points:", points)
print("Average:", average)

Points: [1.5, 0, 0, 0, 0]
Average: 0.3


In [15]:
"""
Proof:

Let

x̄ = (1/n) Σᵢ₌₁ⁿ xᵢ.

Since f is convex, by definition,

f(λx + (1−λ)y)
≤
λf(x) + (1−λ)f(y)

for every 0 ≤ λ ≤ 1.

First consider the case n = 2. Choosing λ = 1/2 gives

f((x₁+x₂)/2)
≤
(f(x₁)+f(x₂))/2,

which proves the lemma for two points.

For the general case, the average

x̄ = (1/n) Σᵢ₌₁ⁿ xᵢ

is a convex combination of the points x₁,…,xₙ with equal weights 1/n.
Applying the convexity inequality repeatedly (or equivalently, Jensen's inequality) yields

f((1/n) Σᵢ₌₁ⁿ xᵢ)
≤
(1/n) Σᵢ₌₁ⁿ f(xᵢ).

Thus, the value of a convex function evaluated at the average of several points is no greater than
the average of the function values, proving Lemma 33.6.
"""

"\nProof:\n\nLet\n\nx̄ = (1/n) Σᵢ₌₁ⁿ xᵢ.\n\nSince f is convex, by definition,\n\nf(λx + (1−λ)y)\n≤\nλf(x) + (1−λ)f(y)\n\nfor every 0 ≤ λ ≤ 1.\n\nFirst consider the case n = 2. Choosing λ = 1/2 gives\n\nf((x₁+x₂)/2)\n≤\n(f(x₁)+f(x₂))/2,\n\nwhich proves the lemma for two points.\n\nFor the general case, the average\n\nx̄ = (1/n) Σᵢ₌₁ⁿ xᵢ\n\nis a convex combination of the points x₁,…,xₙ with equal weights 1/n.\nApplying the convexity inequality repeatedly (or equivalently, Jensen's inequality) yields\n\nf((1/n) Σᵢ₌₁ⁿ xᵢ)\n≤\n(1/n) Σᵢ₌₁ⁿ f(xᵢ).\n\nThus, the value of a convex function evaluated at the average of several points is no greater than\nthe average of the function values, proving Lemma 33.6.\n"

In [16]:
"""
Proof:

Since f is convex,

f((1−λ)x + λy)
≤
(1−λ)f(x) + λf(y)

for every 0 < λ ≤ 1.

Subtract f(x) from both sides:

f((1−λ)x + λy) − f(x)
≤
λ(f(y) − f(x)).

Divide both sides by λ:

[f((1−λ)x + λy) − f(x)] / λ
≤
f(y) − f(x).

Since

(1−λ)x + λy = x + λ(y−x),

the inequality becomes

[f(x + λ(y−x)) − f(x)] / λ
≤
f(y) − f(x).

Taking the limit as λ → 0 and using differentiability,

lim(λ→0)
[f(x + λ(y−x)) − f(x)] / λ
=
∇f(x)ᵀ(y−x).

Therefore,

∇f(x)ᵀ(y−x)
≤
f(y) − f(x),

or equivalently,

f(y)
≥
f(x) + ∇f(x)ᵀ(y−x).

Hence, the tangent line (or tangent hyperplane) of a differentiable convex function is always a global lower bound on the function,
proving Lemma 33.7.
"""

'\nProof:\n\nSince f is convex,\n\nf((1−λ)x + λy)\n≤\n(1−λ)f(x) + λf(y)\n\nfor every 0 < λ ≤ 1.\n\nSubtract f(x) from both sides:\n\nf((1−λ)x + λy) − f(x)\n≤\nλ(f(y) − f(x)).\n\nDivide both sides by λ:\n\n[f((1−λ)x + λy) − f(x)] / λ\n≤\nf(y) − f(x).\n\nSince\n\n(1−λ)x + λy = x + λ(y−x),\n\nthe inequality becomes\n\n[f(x + λ(y−x)) − f(x)] / λ\n≤\nf(y) − f(x).\n\nTaking the limit as λ → 0 and using differentiability,\n\nlim(λ→0)\n[f(x + λ(y−x)) − f(x)] / λ\n=\n∇f(x)ᵀ(y−x).\n\nTherefore,\n\n∇f(x)ᵀ(y−x)\n≤\nf(y) − f(x),\n\nor equivalently,\n\nf(y)\n≥\nf(x) + ∇f(x)ᵀ(y−x).\n\nHence, the tangent line (or tangent hyperplane) of a differentiable convex function is always a global lower bound on the function,\nproving Lemma 33.7.\n'

In [17]:
"""
Proof:

First consider the one-dimensional case. Let

a = x_t,
b = x^r[t+1],
c = x*.

Then

a − c = (a − b) + (b − c).

Squaring both sides gives

(a − c)²
=
(a − b)²
+
2(a − b)(b − c)
+
(b − c)².

Thus,

(x_t − x*)²
=
(x_t − x^r[t+1])²
+
2(x_t − x^r[t+1])(x^r[t+1] − x*)
+
(x^r[t+1] − x*)².

For the general n-dimensional case, let

u = x_t − x^r[t+1]

and

v = x^r[t+1] − x*.

Then

u + v = x_t − x*.

Using the identity

||u + v||²
=
(u + v)^T(u + v)
=
u^Tu + 2u^Tv + v^Tv,

we obtain

||x_t − x*||²
=
||x_t − x^r[t+1]||²
+
2(x_t − x^r[t+1])^T(x^r[t+1] − x*)
+
||x^r[t+1] − x*||².

Therefore, Equation (33.29) follows directly from the expansion of the squared Euclidean norm.
"""

'\nProof:\n\nFirst consider the one-dimensional case. Let\n\na = x_t,\nb = x^r[t+1],\nc = x*.\n\nThen\n\na − c = (a − b) + (b − c).\n\nSquaring both sides gives\n\n(a − c)²\n=\n(a − b)²\n+\n2(a − b)(b − c)\n+\n(b − c)².\n\nThus,\n\n(x_t − x*)²\n=\n(x_t − x^r[t+1])²\n+\n2(x_t − x^r[t+1])(x^r[t+1] − x*)\n+\n(x^r[t+1] − x*)².\n\nFor the general n-dimensional case, let\n\nu = x_t − x^r[t+1]\n\nand\n\nv = x^r[t+1] − x*.\n\nThen\n\nu + v = x_t − x*.\n\nUsing the identity\n\n||u + v||²\n=\n(u + v)^T(u + v)\n=\nu^Tu + 2u^Tv + v^Tv,\n\nwe obtain\n\n||x_t − x*||²\n=\n||x_t − x^r[t+1]||²\n+\n2(x_t − x^r[t+1])^T(x^r[t+1] − x*)\n+\n||x^r[t+1] − x*||².\n\nTherefore, Equation (33.29) follows directly from the expansion of the squared Euclidean norm.\n'

In [18]:
"""
Proof:

The function is

f(w)=w₀+Σⱼ₌₁ⁿwⱼxⱼ,

where x₁,…,xₙ are fixed constants and the variables are
w₀,w₁,…,wₙ.

Let u=(u₀,…,uₙ) and v=(v₀,…,vₙ), and let
0≤λ≤1.

Then

f(λu+(1−λ)v)

= λu₀+(1−λ)v₀
+Σⱼ₌₁ⁿ(λuⱼ+(1−λ)vⱼ)xⱼ

= λ(u₀+Σⱼ₌₁ⁿuⱼxⱼ)
+(1−λ)(v₀+Σⱼ₌₁ⁿvⱼxⱼ)

= λf(u)+(1−λ)f(v).

Therefore,

f(λu+(1−λ)v)
= λf(u)+(1−λ)f(v),

so the convexity inequality holds with equality. Hence, f is a convex function of the variables w₀,w₁,…,wₙ.
In fact, every affine (linear) function is both convex and concave.
"""

'\nProof:\n\nThe function is\n\nf(w)=w₀+Σⱼ₌₁ⁿwⱼxⱼ,\n\nwhere x₁,…,xₙ are fixed constants and the variables are\nw₀,w₁,…,wₙ.\n\nLet u=(u₀,…,uₙ) and v=(v₀,…,vₙ), and let\n0≤λ≤1.\n\nThen\n\nf(λu+(1−λ)v)\n\n= λu₀+(1−λ)v₀\n+Σⱼ₌₁ⁿ(λuⱼ+(1−λ)vⱼ)xⱼ\n\n= λ(u₀+Σⱼ₌₁ⁿuⱼxⱼ)\n+(1−λ)(v₀+Σⱼ₌₁ⁿvⱼxⱼ)\n\n= λf(u)+(1−λ)f(v).\n\nTherefore,\n\nf(λu+(1−λ)v)\n= λf(u)+(1−λ)f(v),\n\nso the convexity inequality holds with equality. Hence, f is a convex function of the variables w₀,w₁,…,wₙ.\nIn fact, every affine (linear) function is both convex and concave.\n'

In [19]:
"""
Prediction → Error → Square the Error → Compute the Gradient → Update the Weights

Example:
Suppose we have one feature (n = 1) and two training examples (m = 2).

Training Data

Example 1:
x^(1) = 2
y^(1) = 5

Example 2:
x^(2) = 4
y^(2) = 9

Suppose our current weights are

w₀ = 0
w₁ = 1

--------------------------------------------------

Step 1: Compute the prediction for each example.

Prediction for Example 1

ŷ^(1)

= w₀ + w₁x^(1)
= 0 + 1(2)
= 2

Prediction for Example 2

ŷ^(2)

= w₀ + w₁x^(2)
= 0 + 1(4)
= 4

--------------------------------------------------

Step 2: Compute the error for each example.

Error^(1)

= ŷ^(1) − y^(1)
= 2 − 5
= -3

Error^(2)

= ŷ^(2) − y^(2)
= 4 − 9
= -5

--------------------------------------------------

Step 3: Compute the objective function.

f(w)

= (-3)² + (-5)²
= 9 + 25
= 34

--------------------------------------------------

Step 4: Compute the gradient with respect to w₀.

∂f/∂w₀

= 2 Σ error
= 2((-3) + (-5))
= 2(-8)
= -16

--------------------------------------------------

Step 5: Compute the gradient with respect to w₁.

∂f/∂w₁

= 2 Σ(error × x)
= 2((-3)(2) + (-5)(4))
= 2(-6 - 20)
= 2(-26)
= -52

--------------------------------------------------

Step 6: Final gradient.

∇f

=

(-16, -52)

--------------------------------------------------

Interpretation:

The gradient tells us the direction in which the error increases the fastest.

Since both components are negative, gradient descent subtracts the gradient, causing both weights to increase.

If the learning rate is η,

w₀ ← w₀ - η(-16) = w₀ + 16η

w₁ ← w₁ - η(-52) = w₁ + 52η

Notice why this makes sense.

Our predictions (2 and 4) are much smaller than the true values (5 and 9), so increasing both the bias (w₀) and
the slope (w₁) will increase future predictions and reduce the squared error.
"""

'\nPrediction → Error → Square the Error → Compute the Gradient → Update the Weights\n\nExample:\nSuppose we have one feature (n = 1) and two training examples (m = 2).\n\nTraining Data\n\nExample 1:\nx^(1) = 2\ny^(1) = 5\n\nExample 2:\nx^(2) = 4\ny^(2) = 9\n\nSuppose our current weights are\n\nw₀ = 0\nw₁ = 1\n\n--------------------------------------------------\n\nStep 1: Compute the prediction for each example.\n\nPrediction for Example 1\n\nŷ^(1)\n\n= w₀ + w₁x^(1)\n= 0 + 1(2)\n= 2\n\nPrediction for Example 2\n\nŷ^(2)\n\n= w₀ + w₁x^(2)\n= 0 + 1(4)\n= 4\n\n--------------------------------------------------\n\nStep 2: Compute the error for each example.\n\nError^(1)\n\n= ŷ^(1) − y^(1)\n= 2 − 5\n= -3\n\nError^(2)\n\n= ŷ^(2) − y^(2)\n= 4 − 9\n= -5\n\n--------------------------------------------------\n\nStep 3: Compute the objective function.\n\nf(w)\n\n= (-3)² + (-5)²\n= 9 + 25\n= 34\n\n--------------------------------------------------\n\nStep 4: Compute the gradient with respect to w₀

In [20]:
"""
Example:

Suppose

w = (3,4),

so

||w||

=

√(3²+4²)

=

5.

Let

x=(2,5)

and

y=(1,1).

The difference between the two points is

x-y=(1,4),

whose length is

||x-y||

=

√17

≈4.12.

The linear function is

f(x)=w₀+w₁x₁+w₂x₂.

Ignoring the constant term w₀ (since it cancels when subtracting),

f(x)-f(y)

=

3(2-1)+4(5-1)

=

3(1)+4(4)

=

19.

The Lipschitz inequality states

|f(x)-f(y)|

≤

||w||||x-y||.

Substituting the values,

19

≤

5√17

≈20.6,

which is true.

Therefore, if the weight vector satisfies

||w||≤B,

then

|f(x)-f(y)|

≤

B||x-y||,

so the Lipschitz constant satisfies

L≤B.

Hence,

L=O(B).
"""

'\nExample:\n\nSuppose\n\nw = (3,4),\n\nso\n\n||w||\n\n=\n\n√(3²+4²)\n\n=\n\n5.\n\nLet\n\nx=(2,5)\n\nand\n\ny=(1,1).\n\nThe difference between the two points is\n\nx-y=(1,4),\n\nwhose length is\n\n||x-y||\n\n=\n\n√17\n\n≈4.12.\n\nThe linear function is\n\nf(x)=w₀+w₁x₁+w₂x₂.\n\nIgnoring the constant term w₀ (since it cancels when subtracting),\n\nf(x)-f(y)\n\n=\n\n3(2-1)+4(5-1)\n\n=\n\n3(1)+4(4)\n\n=\n\n19.\n\nThe Lipschitz inequality states\n\n|f(x)-f(y)|\n\n≤\n\n||w||||x-y||.\n\nSubstituting the values,\n\n19\n\n≤\n\n5√17\n\n≈20.6,\n\nwhich is true.\n\nTherefore, if the weight vector satisfies\n\n||w||≤B,\n\nthen\n\n|f(x)-f(y)|\n\n≤\n\nB||x-y||,\n\nso the Lipschitz constant satisfies\n\nL≤B.\n\nHence,\n\nL=O(B).\n'

In [21]:
"""
Moving torward centroid

Example:

Suppose we have one cluster containing the points

1, 2, and 4,

and the current centroid is

c = 6.

The objective function for this cluster is

f(c)

=

(1-c)² + (2-c)² + (4-c)².

The gradient is

∇f

=

2((c-1)+(c-2)+(c-4)).

Substituting c = 6,

∇f

=

2((6-1)+(6-2)+(6-4))

=

2(5+4+2)

=

22.

Using a learning rate η = 0.1,

the new centroid is

c

←

6 − 0.1(22)

=

3.8.

The centroid moves closer to the points.

On the next iteration,

∇f

=

2((3.8-1)+(3.8-2)+(3.8-4))

=

2(2.8+1.8−0.2)

=

8.8.

Updating again,

c

←

3.8 − 0.1(8.8)

=

2.92.

The centroid continues moving toward the average of the data points. When the gradient becomes zero,

Σ(c−x)=0,

which implies

c=(1/|S|)Σx,

the mean of the cluster. Thus, gradient descent eventually converges to the same centroid that Lloyd's algorithm computes directly by taking the average of the assigned points.
"""

"\nMoving torward centroid\n\nExample:\n\nSuppose we have one cluster containing the points\n\n1, 2, and 4,\n\nand the current centroid is\n\nc = 6.\n\nThe objective function for this cluster is\n\nf(c)\n\n=\n\n(1-c)² + (2-c)² + (4-c)².\n\nThe gradient is\n\n∇f\n\n=\n\n2((c-1)+(c-2)+(c-4)).\n\nSubstituting c = 6,\n\n∇f\n\n=\n\n2((6-1)+(6-2)+(6-4))\n\n=\n\n2(5+4+2)\n\n=\n\n22.\n\nUsing a learning rate η = 0.1,\n\nthe new centroid is\n\nc\n\n←\n\n6 − 0.1(22)\n\n=\n\n3.8.\n\nThe centroid moves closer to the points.\n\nOn the next iteration,\n\n∇f\n\n=\n\n2((3.8-1)+(3.8-2)+(3.8-4))\n\n=\n\n2(2.8+1.8−0.2)\n\n=\n\n8.8.\n\nUpdating again,\n\nc\n\n←\n\n3.8 − 0.1(8.8)\n\n=\n\n2.92.\n\nThe centroid continues moving toward the average of the data points. When the gradient becomes zero,\n\nΣ(c−x)=0,\n\nwhich implies\n\nc=(1/|S|)Σx,\n\nthe mean of the cluster. Thus, gradient descent eventually converges to the same centroid that Lloyd's algorithm computes directly by taking the average of the assig

In [22]:
"""
Key insight: Lloyd's algorithm can be viewed as a specialized form of gradient descent.
Rather than taking many small gradient steps toward the minimum, it computes the point where the gradient
is exactly zero (the cluster mean) in a single update for each cluster, which is why it converges much faster
once the cluster assignments are fixed.
"""

"\nKey insight: Lloyd's algorithm can be viewed as a specialized form of gradient descent.\nRather than taking many small gradient steps toward the minimum, it computes the point where the gradient\nis exactly zero (the cluster mean) in a single update for each cluster, which is why it converges much faster\nonce the cluster assignments are fixed.\n"

# Newton's Method vs. Gradient Descent

Newton's method is an iterative algorithm used to find the **root** of a function, that is, a value $x^*$ such that

$$
f(x^*) = 0.
$$

Unlike gradient descent, which minimizes a function by moving in the direction of the negative gradient, Newton's method approximates the function locally with its tangent line and repeatedly jumps to where that tangent line crosses the x-axis.

---

## (a) Deriving Newton's Update Rule

Suppose we are currently at the point $x^{(t)}$.

The equation of the tangent line to $f(x)$ at this point is

$$
y = f'(x^{(t)})(x-x^{(t)}) + f(x^{(t)}).
$$

Newton's method chooses the x-intercept of this tangent line as the next approximation. Since the x-intercept occurs when $y=0$, we set

$$
0 = f'(x^{(t)})(x-x^{(t)}) + f(x^{(t)}).
$$

Solving for $x$ gives

$$
x-x^{(t)}
=
-\frac{f(x^{(t)})}{f'(x^{(t)})},
$$

or equivalently,

$$
\boxed{
x^{(t+1)}
=
x^{(t)}
-
\frac{f(x^{(t)})}{f'(x^{(t)})}
}
$$

This is the Newton update rule.

The intuition is simple: rather than taking a small step downhill like gradient descent, Newton's method asks, "If the tangent line were the entire function, where would its root be?" It then jumps directly to that point.

---

## (b) Error Analysis Using Taylor Expansion

Suppose the function has exactly one root $x^*$, and define the error at iteration $t$ as

$$
\epsilon^{(t)}
=
|x^{(t)}-x^*|.
$$

Since

$$
f(x^*)=0,
$$

we apply the Taylor expansion of $f(x^*)$ around the current approximation $x^{(t)}$:

$$
0
=
f(x^{(t)})
+
f'(x^{(t)})(x^*-x^{(t)})
+
\frac12
f''(y^{(t)})
(x^*-x^{(t)})^2,
$$

where $y^{(t)}$ lies between $x^{(t)}$ and $x^*$.

Substituting this expression into Newton's update rule and simplifying yields

$$
x^{(t+1)}-x^*
=
\frac{f''(y^{(t)})}
{2f'(x^{(t)})}
(x^*-x^{(t)})^2.
$$

Taking absolute values gives

$$
\boxed{
\epsilon^{(t+1)}
=
\frac{|f''(y^{(t)})|}
{2|f'(y^{(t)})|}
(\epsilon^{(t)})^2.
}
$$

This equation is extremely important because it describes how the error changes from one iteration to the next.

Notice that the new error depends on the **square** of the previous error.

---

## (c) Why Newton's Method is Quadratically Convergent

Suppose

$$
\frac{|f''(y^{(t)})|}
{2|f'(y^{(t)})|}
\le c
$$

for some constant $c$, and suppose the initial error satisfies

$$
\epsilon^{(0)}<1.
$$

Then

$$
\epsilon^{(t+1)}
\le
c(\epsilon^{(t)})^2.
$$

Expanding the first few iterations,

$$
\epsilon^{(1)}
\le
c\epsilon_0^2,
$$

$$
\epsilon^{(2)}
\le
c^3\epsilon_0^4,
$$

$$
\epsilon^{(3)}
\le
c^7\epsilon_0^8.
$$

The exponent on the error doubles every iteration.

In general,

$$
\epsilon^{(t)}
\le
c^{2^t-1}
(\epsilon^{(0)})^{2^t}.
$$

To achieve an error smaller than a desired tolerance $\Delta$, we require

$$
\epsilon^{(t)}
\le
\Delta.
$$

Solving for the number of iterations shows

$$
\boxed{
t
=
O\left(\log\log\frac1\Delta\right).
}
$$

This is called **quadratic convergence**.

The name comes from the fact that the error is squared at every iteration.

For example, if the error is

- $10^{-2}$,
- after one iteration it becomes approximately $10^{-4}$,
- then $10^{-8}$,
- then $10^{-16}$.

Each iteration roughly doubles the number of correct digits.

For comparison, gradient descent has linear convergence and typically requires

$$
O\left(\log\frac1\Delta\right)
$$

iterations, which is much slower.

---

## (d) Example: $f(x)=(x-3)^2$

Consider the function

$$
f(x)=(x-3)^2.
$$

This function has

- a minimum at $x=3$,
- and a root at $x=3$.

Assume the starting point is

$$
x^{(0)}=3.5.
$$

### Gradient Descent

The derivative is

$$
f'(x)=2(x-3).
$$

Using gradient descent,

$$
x^{(t+1)}
=
x^{(t)}
-
\eta f'(x^{(t)}).
$$

Choosing the common learning rate

$$
\eta=\frac12,
$$

gives

$$
x^{(t+1)}
=
x^{(t)}
-
(x^{(t)}-3)
=
3.
$$

Starting from

$$
x^{(0)}=3.5,
$$

we immediately obtain

$$
x^{(1)}=3.
$$

Therefore, gradient descent reaches the minimum in

$$
\boxed{1\text{ iteration}.}
$$

---

### Newton's Method

Newton's update is

$$
x^{(t+1)}
=
x^{(t)}
-
\frac{(x^{(t)}-3)^2}
{2(x^{(t)}-3)}
=
\frac{x^{(t)}+3}{2}.
$$

Starting from 3.5,

| Iteration | Approximation | Error |
|-----------|--------------|-------|
| 0 | 3.5000 | 0.5000 |
| 1 | 3.2500 | 0.2500 |
| 2 | 3.1250 | 0.1250 |
| 3 | 3.0625 | 0.0625 |
| 4 | 3.03125 | 0.03125 |

The error is cut in half every iteration.

Interestingly, Newton's method is **not quadratically convergent** for this function.

The reason is that quadratic convergence requires

$$
f'(x^*)\neq0.
$$

Here,

$$
f'(3)=0,
$$

so the root has multiplicity two (it is a repeated root).

Because the denominator of Newton's update becomes zero at the root, the assumptions behind the quadratic convergence proof no longer hold. Instead of squaring the error each iteration, Newton's method only halves it, resulting in **linear convergence**.

---

## Key Takeaways

- Gradient descent minimizes a function by following the negative gradient.
- Newton's method finds a root by repeatedly moving to the x-intercept of the tangent line.
- Newton's update rule is

$$
x^{(t+1)}
=
x^{(t)}
-
\frac{f(x^{(t)})}{f'(x^{(t)})}.
$$

- Under the assumptions that $f'(x^*)\neq0$ and $f''$ is continuous, Newton's method has **quadratic convergence**:

$$
\epsilon^{(t+1)}
=
O((\epsilon^{(t)})^2).
$$

- Therefore, Newton's method requires only

$$
O\left(\log\log\frac1\Delta\right)
$$

iterations to reach an accuracy of $\Delta$.

- However, if the root is repeated (as in $(x-3)^2$), then $f'(x^*)=0$, quadratic convergence is lost, and Newton's method converges only linearly.

- For the function $(x-3)^2$ starting at $x=3.5$, gradient descent with $\eta=\frac12$ reaches the minimum in one iteration, while Newton's method only cuts the error in half each iteration.

In [23]:
"""
The potential function is still the total weight Z^(t)

If Hedge is likely to make a mistake, a large fraction of the total weight belongs to incorrect experts, so the total weight shrinks significantly.
The best expert loses weight only when it actually makes a mistake.
Comparing these two facts bounds Hedge's expected mistakes.

The only substantial difference from Weighted Majority is that Hedge is randomized, so instead of counting actual mistakes,
we analyze the expected mistake probability P_t at each round. This is why the final guarantee
is an expected mistake bound rather than a deterministic one.
"""

"\nThe potential function is still the total weight Z^(t)\n\nIf Hedge is likely to make a mistake, a large fraction of the total weight belongs to incorrect experts, so the total weight shrinks significantly.\nThe best expert loses weight only when it actually makes a mistake.\nComparing these two facts bounds Hedge's expected mistakes.\n\nThe only substantial difference from Weighted Majority is that Hedge is randomized, so instead of counting actual mistakes,\nwe analyze the expected mistake probability P_t at each round. This is why the final guarantee\nis an expected mistake bound rather than a deterministic one.\n"

# Hedge Algorithm Mistake Bound

The Hedge algorithm is a randomized version of the Weighted Majority algorithm.

Unlike Weighted Majority, which always predicts according to the expert with the largest weight, Hedge chooses an expert randomly.

At round $t$, expert $i$ is selected with probability

$$
p_i^{(t)}
=
\frac{w_i^{(t)}}{Z^{(t)}},
$$

where

$$
Z^{(t)}
=
\sum_{i=1}^{n}w_i^{(t)}.
$$

If the selected expert makes a mistake, then Hedge also makes a mistake.

The update rule is

$$
w_i^{(t+1)}
=
w_i^{(t)}e^{-\epsilon},
$$

for every expert that made a mistake, where

$$
0<\epsilon<1.
$$

Correct experts keep their current weight.

---

## Step 1. Expected Mistake Probability

Let

$$
M^{(t)}
$$

be the set of experts that make a mistake during round $t$.

Since Hedge selects experts according to their normalized weights, the probability that Hedge makes a mistake is

$$
P_t
=
\sum_{i\in M^{(t)}}
\frac{w_i^{(t)}}{Z^{(t)}}.
$$

Therefore,

$$
P_t
$$

is exactly the expected number of mistakes made during round $t$.

The expected total number of mistakes over all $T$ rounds is

$$
\sum_{t=1}^{T}P_t.
$$

---

## Step 2. Analyze the Total Weight

Initially,

$$
w_i^{(1)}=1
$$

for every expert, so

$$
Z^{(1)}=n.
$$

After one update,

$$
Z^{(t+1)}
=
\sum_{i\notin M^{(t)}}w_i^{(t)}
+
e^{-\epsilon}
\sum_{i\in M^{(t)}}w_i^{(t)}.
$$

Factor out $Z^{(t)}$:

$$
Z^{(t+1)}
=
Z^{(t)}
\left(
1-(1-e^{-\epsilon})P_t
\right).
$$

Using the inequality

$$
1-x\le e^{-x},
$$

we obtain

$$
Z^{(t+1)}
\le
Z^{(t)}
e^{-(1-e^{-\epsilon})P_t}.
$$

Applying this repeatedly over all rounds,

$$
Z^{(T+1)}
\le
n
\exp
\left(
-(1-e^{-\epsilon})
\sum_{t=1}^{T}P_t
\right).
$$

---

## Step 3. Lower Bound Using the Best Expert

Suppose the best expert makes

$$
m^*
$$

mistakes.

Every mistake multiplies that expert's weight by

$$
e^{-\epsilon},
$$

so after $T$ rounds,

$$
w_{\text{best}}
=
e^{-\epsilon m^*}.
$$

Since the total weight must be at least the weight of the best expert,

$$
Z^{(T+1)}
\ge
e^{-\epsilon m^*}.
$$

---

## Step 4. Combine the Bounds

Since

$$
e^{-\epsilon m^*}
\le
n
\exp
\left(
-(1-e^{-\epsilon})
\sum P_t
\right),
$$

take logarithms:

$$
-\epsilon m^*
\le
\ln n
-
(1-e^{-\epsilon})
\sum P_t.
$$

Rearranging,

$$
\sum P_t
\le
\frac{\ln n+\epsilon m^*}
{1-e^{-\epsilon}}.
$$

---

## Step 5. Simplify the Denominator

Using the inequality

$$
e^{-\epsilon}
\le
1-\epsilon+\frac{\epsilon^2}{2},
$$

we obtain

$$
1-e^{-\epsilon}
\ge
\epsilon-\frac{\epsilon^2}{2}.
$$

Since

$$
\frac1{1-\epsilon/2}
\le
1+\epsilon,
$$

it follows that

$$
\frac1{1-e^{-\epsilon}}
\le
\frac1{\epsilon}+\!1.
$$

Substituting into the previous inequality gives

$$
\sum P_t
\le
m^*
+
\frac{\ln n}{\epsilon}
+
\epsilon T.
$$

Since

$$
\sum P_t
$$

is exactly the expected number of mistakes made by Hedge,

$$
\boxed{
\mathbb{E}[\text{Mistakes}]
\le
m^*
+
\frac{\ln n}{\epsilon}
+
\epsilon T.
}
$$

---

## Intuition

The proof follows the same "potential function" argument as the Weighted Majority algorithm.

The potential function is the total weight

$$
Z^{(t)}=\sum_i w_i^{(t)}.
$$

Every time Hedge is likely to make a mistake, a larger fraction of the total weight belongs to incorrect experts, causing the total weight to shrink.

On the other hand, the best expert loses weight only when it actually makes a mistake.

Comparing the upper bound on the total weight with the lower bound given by the best expert produces a bound on the expected number of mistakes.

Unlike Weighted Majority, Hedge is randomized, so the analysis bounds the **expected** number of mistakes rather than the exact number. The final guarantee shows that Hedge performs almost as well as the best expert, differing by only an additive term of

$$
\frac{\ln n}{\epsilon}+\epsilon T.
$$

# Lloyd's Procedure Does Not Always Find the Global Optimum (Even in One Dimension)

Lloyd's procedure (the standard k-means algorithm) is an iterative algorithm for solving the k-means clustering problem. At each iteration, it performs two steps:

1. **Assignment Step:** Assign every point to the nearest cluster center.
2. **Update Step:** Replace each cluster center with the mean of the points assigned to that cluster.

Although Lloyd's procedure is guaranteed to decrease (or leave unchanged) the objective function at every iteration and eventually terminate, it is **not guaranteed to find the global optimum**. Instead, it may converge to a **local optimum** that depends on the initial placement of the cluster centers.

---

## Counterexample

Consider the one-dimensional set of points

$$
S=\{1,2,4,5,9,10\}
$$

and suppose we wish to find

$$
k=2
$$

clusters.

Choose the initial centers to be

$$
4 \quad \text{and} \quad 5.
$$

Because both centers begin near the left side of the data, Lloyd's procedure converges to the clustering

$$
C_1=\{1,2,4\},
$$

$$
C_2=\{5,9,10\}.
$$

The corresponding cluster centers are

$$
\mu_1=\frac{1+2+4}{3}=\frac73\approx2.33,
$$

$$
\mu_2=\frac{5+9+10}{3}=8.
$$

At this point, reassigning the points does not change either cluster, so Lloyd's procedure terminates.

---

## Objective Function of Lloyd's Solution

The k-means objective function is

$$
f(S,C)
=
\sum_{x\in S}
(x-\mu(x))^2,
$$

where $\mu(x)$ is the center of the cluster containing point $x$.

For the first cluster,

$$
(1-2.33)^2+(2-2.33)^2+(4-2.33)^2
=
1.78+0.11+2.78
=
4.67.
$$

For the second cluster,

$$
(5-8)^2+(9-8)^2+(10-8)^2
=
9+1+4
=
14.
$$

Therefore,

$$
f(S,C)
=
4.67+14
=
18.67.
$$

---

## A Better Clustering

Now consider the clustering

$$
C_1=\{1,2,4,5\},
$$

$$
C_2=\{9,10\}.
$$

The centers are

$$
\mu_1=\frac{1+2+4+5}{4}=3,
$$

$$
\mu_2=\frac{9+10}{2}=9.5.
$$

Its objective function is

First cluster:

$$
(1-3)^2+(2-3)^2+(4-3)^2+(5-3)^2
=
4+1+1+4
=
10.
$$

Second cluster:

$$
(9-9.5)^2+(10-9.5)^2
=
0.25+0.25
=
0.5.
$$

Thus,

$$
f(S,C)
=
10.5.
$$

Since

$$
10.5<18.67,
$$

this clustering is strictly better than the one returned by Lloyd's procedure.

---

## Conclusion

Lloyd's procedure terminates because no point changes clusters after the final update. However, the clustering it returns does **not** minimize the k-means objective function.

Therefore,

$$
\boxed{\text{Lloyd's procedure can converge to a local optimum even in one dimension.}}
$$

This example demonstrates that the final clustering produced by Lloyd's algorithm depends on the initial placement of the cluster centers. Although the objective function decreases at every iteration, the algorithm is **not guaranteed to find the globally optimal clustering**.

In [24]:
import random

def stochastic_gradient_descent(x, y, learning_rate=0.01, epochs=100):
    a = 0.0
    b = 0.0
    n = len(x)

    for epoch in range(epochs):
        # Pick one random point
        i = random.randint(0, n - 1)

        prediction = a * x[i] + b
        error = prediction - y[i]

        # SGD update
        a -= learning_rate * error * x[i]
        b -= learning_rate * error

    return a, b

# Example

x = [1,2,3,4,5]
y = [2,4,6,8,10]

a, b = stochastic_gradient_descent(x, y)

print("Slope:", a)
print("Intercept:", b)

Slope: 1.8895870305154365
Intercept: 0.3896727323596746


In [25]:
"""
Gradient descent computes the exact gradient each iteration, but that requires looking at all T training examples.

SGD looks at only one randomly selected example, making each update much cheaper.

Although every update is noisy, the expected update points in exactly the same direction as the true gradient because the sample is chosen uniformly at random.

The randomness averages out over many iterations, so SGD converges to the optimum in expectation.
"""

'\nGradient descent computes the exact gradient each iteration, but that requires looking at all T training examples.\n\nSGD looks at only one randomly selected example, making each update much cheaper.\n\nAlthough every update is noisy, the expected update points in exactly the same direction as the true gradient because the sample is chosen uniformly at random.\n\nThe randomness averages out over many iterations, so SGD converges to the optimum in expectation.\n'

# Stochastic Gradient Descent (SGD) for Linear Regression

Stochastic Gradient Descent (SGD) is a variant of gradient descent used when the training data is not available all at once or when the dataset is very large. Instead of computing the gradient using every training example, SGD updates the model parameters using **one randomly selected training example at a time**. This makes each update much faster and allows the algorithm to learn in an online setting where data arrives sequentially.

---

## Problem Setup

Suppose we want to fit the line

$$
f(x)=ax+b
$$

to a set of training examples

$$
S=\{(x_1,y_1),(x_2,y_2),\ldots,(x_T,y_T)\}.
$$

The goal is to choose the parameters \(a\) (the slope) and \(b\) (the intercept) that minimize the least-squares error.

The least-squares objective function is

$$
J(a,b)
=
\frac12
\sum_{i=1}^{T}
(ax_i+b-y_i)^2.
$$

In ordinary gradient descent, every iteration computes the gradient using **all** \(T\) training examples.

In stochastic gradient descent, however, one training example is chosen uniformly at random during each iteration, and only that example is used to update the parameters.

---

## Loss Function for One Training Example

Suppose the randomly selected training example is

$$
(x_i,y_i).
$$

The loss associated with this single example is

$$
\ell_i(a,b)
=
\frac12(ax_i+b-y_i)^2.
$$

Define the prediction error as

$$
e_i=ax_i+b-y_i.
$$

This quantity measures how far the current prediction is from the true value.

---

## Compute the Gradient

Differentiate the loss with respect to each parameter.

The derivative with respect to the slope is

$$
\frac{\partial \ell_i}{\partial a}
=
(ax_i+b-y_i)x_i.
$$

The derivative with respect to the intercept is

$$
\frac{\partial \ell_i}{\partial b}
=
ax_i+b-y_i.
$$

These derivatives form a stochastic estimate of the true gradient because they are computed from only one randomly selected training example.

---

## SGD Update Rule

Using a learning rate \(\eta\), the parameters are updated as

$$
a
\leftarrow
a-\eta(ax_i+b-y_i)x_i,
$$

$$
b
\leftarrow
b-\eta(ax_i+b-y_i).
$$

These updates are repeated many times, each time selecting a different random training example.

Unlike ordinary gradient descent, which performs one expensive update using the entire dataset, SGD performs many inexpensive updates using individual samples.

---

## Python Implementation

```python
import random

def stochastic_gradient_descent(x, y, learning_rate=0.01, epochs=1000):

    # Initialize parameters
    a = 0.0
    b = 0.0

    n = len(x)

    for epoch in range(epochs):

        # Randomly choose one training example
        i = random.randint(0, n - 1)

        prediction = a * x[i] + b
        error = prediction - y[i]

        # Update parameters
        a -= learning_rate * error * x[i]
        b -= learning_rate * error

    return a, b


# Example

x = [1, 2, 3, 4, 5]
y = [2, 4, 6, 8, 10]

a, b = stochastic_gradient_descent(x, y)

print("Slope:", a)
print("Intercept:", b)
```

---

## Why This Works

Ordinary gradient descent computes the exact gradient

$$
\nabla J
=
\sum_{i=1}^{T}
\nabla \ell_i.
$$

Stochastic gradient descent computes the gradient using only one randomly selected example,

$$
\nabla \ell_i.
$$

Because each training example is selected uniformly at random,

$$
E[\nabla \ell_i]
=
\frac1T
\sum_{i=1}^{T}
\nabla \ell_i
=
\nabla J.
$$

Therefore, the stochastic gradient is an **unbiased estimate** of the true gradient.

Although each update is noisy, the average update points in the same direction as the true gradient.

---

## Expected Error Analysis

The analysis of SGD closely follows the proof for ordinary gradient descent.

For ordinary gradient descent, the optimization error after \(T\) iterations satisfies

$$
f(\bar{w})-f(w^*)
\le
\frac{LR^2}{2T},
$$

where

- \(L\) is an upper bound on the gradient,
- \(R\) is an upper bound on the distance from the optimum,
- \(T\) is the number of iterations.

Since the stochastic gradient is an unbiased estimator of the true gradient,

$$
E[g_t]
=
\nabla f(w_t),
$$

taking expectations throughout the gradient descent proof gives

$$
E[f(\bar{w})]-f(w^*)
\le
\frac{LR^2}{2T}.
$$

Therefore, the expected optimization error satisfies

$$
\boxed{
E[\text{Error}]
=
O\left(\frac{LR^2}{T}\right).
}
$$

---

## Intuition

Gradient descent computes the exact gradient using every training example before taking a step toward the optimum. This produces accurate updates but can be computationally expensive for large datasets.

Stochastic Gradient Descent instead computes the gradient using a single randomly selected training example. Each update is therefore much cheaper, although it introduces randomness into the optimization process.

The key insight is that the randomness averages out over many iterations. Since the stochastic gradient is an unbiased estimate of the true gradient, the algorithm still moves toward the optimum on average.

As the number of iterations increases, SGD converges to the optimal solution **in expectation**, and its expected optimization error decreases according to

$$
\boxed{
E[\text{Error}]
=
O\left(\frac{LR^2}{T}\right).
}
$$

This makes SGD one of the most widely used optimization algorithms in machine learning, especially for large datasets and online learning problems where training examples arrive one at a time.